fabtests/unit/getinfo_test.c

```diff
@@ -964,6 +964,259 @@ getinfo_test(caps, 4, "Test for capability bit regression",
+		FI_LOCAL_COMM | FI_REMOTE_COMM;
+	int ret;
+
+	/* Verify that the provider meets the required capabilities of this
+	 * test.
+	 */
+	hints->caps = required_caps;
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_DEBUG("No available providers supporting FI_MSG | FI_RMA\n");
+		goto out;
+	}
+
+	hints->domain_attr->name = strdup((*info)->domain_attr->name);
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	if (!hints->domain_attr->name) {
+		ret = -FI_ENOMEM;
+		goto out;
+	}
+
+	/* Pass in the majority of the require test capabilities as optional.
+	 * This capabilites should be set in the return fi_info::caps. In
+	 * addition, since user hint domain_attr::caps, tx_attr::caps, and
+	 * rx_attr::caps are zero, these returns caps should be set based on
+	 * fi_info::caps
+	 */
+	hints->caps = FI_TAGGED;
+	hints->optional_caps = FI_MSG | FI_MULTI_RECV | FI_LOCAL_COMM |
+		FI_REMOTE_COMM;
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_DEBUG("fi_getinfo with optional_caps failed to return\n");
+		goto out;
+	}
+
+	ret = test_valid_optional_cap_info(*info, required_caps,
+					   required_caps & DOMAIN_CAPS,
+					   required_caps & (PRIMARY_TX_CAPS | OPT_TX_CAPS),
+					   required_caps & (PRIMARY_RX_CAPS | OPT_RX_CAPS));
+	if (ret != FI_SUCCESS)
+		return ret;
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	/* Verify that using domain_attr::optional_caps, tx_attr::optional_caps,
+	 * and rx_attr::optional_caps works as expected.
+	 */
+	hints->caps = required_caps;
+	hints->optional_caps = 0;
+	hints->domain_attr->caps = FI_REMOTE_COMM;
+	hints->tx_attr->caps = FI_TAGGED;
+	hints->rx_attr->caps = FI_TAGGED;
+
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_DEBUG("fi_getinfo with failed to return\n");
+		goto out;
+	}
+
+	ret = test_valid_optional_cap_info(*info, required_caps, FI_REMOTE_COMM,
+					   FI_TAGGED, FI_TAGGED);
+	if (ret != FI_SUCCESS)
+		return ret;
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	hints->domain_attr->optional_caps = FI_LOCAL_COMM;
+	hints->tx_attr->optional_caps = FI_MSG | FI_MULTI_RECV;
+	hints->rx_attr->optional_caps = FI_MSG | FI_MULTI_RECV;
+
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_DEBUG("fi_getinfo with optional_caps failed to return\n");
+		goto out;
+	}
+
+	ret = test_valid_optional_cap_info(*info, required_caps,
+					   required_caps & DOMAIN_CAPS,
+					   required_caps & (PRIMARY_TX_CAPS | OPT_TX_CAPS),
+					   required_caps & (PRIMARY_RX_CAPS | OPT_RX_CAPS));
+	if (ret != FI_SUCCESS)
+		return ret;
+
+out:
+	return ret;
+}
+
+static int test_ignore_optional_cap(char *node, char *service, uint64_t flags,
+				    struct fi_info *hints,
+				    struct fi_info **info)
+{
+	uint64_t required_caps = FI_MSG | FI_TAGGED | FI_MULTI_RECV |
+		FI_LOCAL_COMM | FI_REMOTE_COMM;
```

Comment:
FI_MULTI_RECV, FI_LOCAL_COMM, FI_REMOTE_COMM are secondary caps and can always be set, even if not requested.  When setting the optional_caps, we'll want to specify primary caps only.  Maybe use FI_COLLECTIVE and FI_MULTICAST there for bits that are unlikely to be supported by the provider, and FI_RMA as a bit that is likely.

Reply:
Will fix.